### ISSUE TITLE
When DNS resolution fails, I get a "TypeError: an integer is required (got type bytes)" when calling monkey-patched `getaddrinfo`

### ISSUE THREAD

**Issue by jeffcasavant** [gevent/gevent#1491]
* gevent version: ~1.4.0~
* Python version: ~cPython 3.7.5 installed via Brew~ cPython 3.8 inside the `python:3-alpine` container
* Operating System: Alpine Linux Docker container based on `python:3-alpine` running on MacOS Catalina - `uname -a`: Darwin jcasavant-MacBook-Pro 19.0.0 Darwin Kernel Version 19.0.0: Thu Oct 17 16:17:15 PDT 2019; root:xnu-6153.41.3~29/RELEASE_X86_64 x86_64

### Description:

I had my Docker daemon configured with Google public DNS to work around a separate issue I was experiencing.  When returning to a different project which uses an internal LDAP service for authentication, I was unable to run the project locally.  I received the following traceback:

```python
2019-12-10 18:32:54,122 - lib.auth - DEBUG - Connect to LDAP server ldaps://internal-server:636
Traceback (most recent call last):
  File "./app.py", line 206, in <module>
    main(main_module=True)
  File "./app.py", line 126, in main
    from lib import resources
  File "/app/lib/resources.py", line 51, in <module>
    AUTH = UserAuth(CONFIG)
  File "/app/util/singleton.py", line 20, in __init__
    Singletonized.__instance = internal_class(*args, **kwargs)
  File "/app/lib/auth.py", line 105, in __init__
    self._recreate_conn()
  File "/app/lib/auth.py", line 112, in _recreate_conn
    self._conn_real()
  File "/app/lib/auth.py", line 117, in _conn_real
    self.conn  = Connection(self.server, user=self.config.get('ldap.bind'),
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/connection.py", line 326, in __init__
    self.do_auto_bind()
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/connection.py", line 341, in do_auto_bind
    self.open(read_server_info=False)
  File "/usr/local/lib/python3.8/site-packages/ldap3/strategy/sync.py", line 56, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.8/site-packages/ldap3/strategy/base.py", line 123, in open
    for candidate_address in self.connection.server.candidate_addresses():
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/server.py", line 546, in candidate_addresses
    for address in self.address_info:
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/server.py", line 246, in address_info
    addresses = socket.getaddrinfo(self.host, self.port, socket.AF_UNSPEC, socket.SOCK_STREAM, socket.IPPROTO_TCP, socket.AI_ADDRCONFIG | socket.AI_V4MAPPED)
  File "/usr/local/lib/python3.8/site-packages/gevent/_socketcommon.py", line 212, in getaddrinfo
    addrlist = get_hub().resolver.getaddrinfo(host, port, family, type, proto, flags)
  File "/usr/local/lib/python3.8/site-packages/gevent/resolver/thread.py", line 65, in getaddrinfo
    return self.pool.apply(_socket.getaddrinfo, args, kwargs)
  File "/usr/local/lib/python3.8/site-packages/gevent/pool.py", line 159, in apply
    return self.spawn(func, *args, **kwds).get()
  File "src/gevent/event.py", line 268, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 296, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 286, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 266, in gevent._event.AsyncResult._raise_exception
  File "src/gevent/event.py", line 211, in gevent._event.AsyncResult.exc_info.__get__
  File "/usr/local/lib/python3.8/site-packages/gevent/_tblib.py", line 371, in g
    return f(a)
  File "/usr/local/lib/python3.8/site-packages/gevent/_tblib.py", line 432, in load_traceback
    return loads(s)
  File "/usr/local/lib/python3.8/site-packages/gevent/_tblib.py", line 337, in unpickle_traceback
    return ret.as_traceback()
  File "/usr/local/lib/python3.8/site-packages/gevent/_tblib.py", line 202, in as_traceback
    code = CodeType(
TypeError: an integer is required (got type bytes)
make: *** [run-dev] Error 1
```

When I used `importlib.reload(socket)` to un-monkey-patch the socket library, I received the following more useful error:

```python
2019-12-10 18:34:04,090 - lib.auth - DEBUG - Connect to LDAP server ldaps://internal-server:636
Traceback (most recent call last):
  File "./app.py", line 209, in <module>
    main(main_module=True)
  File "./app.py", line 129, in main
    from lib import resources
  File "/app/lib/resources.py", line 51, in <module>
    AUTH = UserAuth(CONFIG)
  File "/app/util/singleton.py", line 20, in __init__
    Singletonized.__instance = internal_class(*args, **kwargs)
  File "/app/lib/auth.py", line 105, in __init__
    self._recreate_conn()
  File "/app/lib/auth.py", line 112, in _recreate_conn
    self._conn_real()
  File "/app/lib/auth.py", line 117, in _conn_real
    self.conn  = Connection(self.server, user=self.config.get('ldap.bind'),
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/connection.py", line 326, in __init__
    self.do_auto_bind()
  File "/usr/local/lib/python3.8/site-packages/ldap3/core/connection.py", line 341, in do_auto_bind
    self.open(read_server_info=False)
  File "/usr/local/lib/python3.8/site-packages/ldap3/strategy/sync.py", line 56, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.8/site-packages/ldap3/strategy/base.py", line 150, in open
    raise LDAPSocketOpenError('invalid server address')
ldap3.core.exceptions.LDAPSocketOpenError: invalid server address
make: *** [run-dev] Error 1
```

`gevent` seems to report the wrong error back when failing to resolve an address.

### What I've run:

Here's my basic monkey-patch setup at the top of my main file.

```python
from gevent import monkey
monkey.patch_all()
```

And with `socket` being reloaded:

```python
import socket
from gevent import monkey
monkey.patch_all()
from importlib import reload
reload(socket)
```

The actual call to getaddrinfo can be seen in the traceback above.  I'm not in control of it since I'm using python-ldap3 to connect to an LDAP server, and it's calling into the gevent-patched version of socket.  I don't think my creation of an `ldap3` `Connection` object will be helpful to you since it's a call into someone else's library, so I did not include it here.

**Comment by dbanty** [comment]
I'm seeing something very similar.  Using gevent under flask-socketio, if I pass "0.0.0.0" in for my host, running fails with a similar exception which I'll include below.  Interestingly this only fails for me with Python 3.8.0, not 3.7.5, and only on macOS.  I have debian-based containers successfully running the same code with 3.8.0.

```
Traceback (most recent call last):
  File "/Users/dylan/.pyenv/versions/3.8.0/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/dylan/.pyenv/versions/3.8.0/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/dylan/projects/cloud/simc-server/spotr_server/__main__.py", line 63, in <module>
    socketio.run(
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/flask_socketio/__init__.py", line 602, in run
    self.wsgi_server.serve_forever()
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/baseserver.py", line 367, in serve_forever
    self.start()
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/baseserver.py", line 305, in start
    self.init_socket()
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/pywsgi.py", line 1491, in init_socket
    self.update_environ()
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/pywsgi.py", line 1503, in update_environ
    name = socket.getfqdn(address[0])
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_socketcommon.py", line 269, in getfqdn
    hostname, aliases, _ = gethostbyaddr(name)
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_socketcommon.py", line 241, in gethostbyaddr
    return get_hub().resolver.gethostbyaddr(ip_address)
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/resolver/thread.py", line 68, in gethostbyaddr
    return self.pool.apply(_socket.gethostbyaddr, args, kwargs)
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/pool.py", line 159, in apply
    return self.spawn(func, *args, **kwds).get()
  File "src/gevent/event.py", line 268, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 296, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 286, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 266, in gevent._event.AsyncResult._raise_exception
  File "src/gevent/event.py", line 211, in gevent._event.AsyncResult.exc_info.__get__
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_tblib.py", line 371, in g
    return f(a)
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_tblib.py", line 432, in load_traceback
    return loads(s)
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_tblib.py", line 337, in unpickle_traceback
    return ret.as_traceback()
  File "/Users/dylan/projects/cloud/simc-server/.venv/lib/python3.8/site-packages/gevent/_tblib.py", line 202, in as_traceback
    code = CodeType(
TypeError: an integer is required (got type bytes)

Process finished with exit code 1
```

**Comment by jamadden** [comment]
(@jeffcasavant the tracebacks you posted show Python 3.8, not 3.7, so I'm going to assume that's what you meant.)

gevent 1.4 is not compatible with Python 3.8. Please use gevent >= 1.5a2 on Python 3.8. This is one of the (several) known issues of running gevent 1.4 on Python 3.8.

**Comment by jeffcasavant** [comment]
Ha, my fault.  I was running inside a python3.8 docker container and gave you my host's info.  I'll see if I can replicate this with the newer version (and a bit more attention to detail).

**Comment by jeffcasavant** [comment]
Yeah, this is fixed in 1.5a2.  Closing this issue.
